### ISSUE TITLE
Inconsistent behavior of .diag() method for different Lazy Tensors in batch mode

### ISSUE THREAD

**Issue by Balandat** [cornellius-gp/gpytorch#292]
E.g: 

```
> cv.lazy_vars
[<gpytorch.lazy.lazy_evaluated_kernel_tensor.LazyEvaluatedKernelTensor at 0x7f6139087278>,
 <gpytorch.lazy.matmul_lazy_tensor.MatmulLazyTensor at 0x7f6139087780>]

> [lv.size() for lv in cv.lazy_vars]
[torch.Size([2, 1, 1]), torch.Size([2, 1, 1])]

>diags = [lazy_var.diag().contiguous() for lazy_var in cv.lazy_vars]
>[diag.size() for diag in diags]
[torch.Size([2]), torch.Size([2, 1])]
```

This causes inconsistencies when using batch mode. My suggestion is to ensure that all `.diag()` calls reduce the last two dimensions of the tensor, i.e. for a `b1 x b2 x n x n` LazyTensor this would return a Tensor of Size `b1 x b2 x n`. This is the same as the behavior of `torch.diagonal(X, dim1=-2, dim2=-1)`.

**Comment by gpleiss** [comment]
This will probably be fixed with the kernel refactor I'm working on

**Comment by gpleiss** [comment]
I'm going to call this closed by #293
